module-audio/board/rt1051/SAIAudioDevice.cpp

```diff
@@ -97,3 +100,18 @@ void SAIAudioDevice::disableOutput()
 {
     txEnabled = false;
 }
+AudioDevice::RetCode SAIAudioDevice::setOutputVolume(float vol)
+{
+    vol = std::clamp(vol, minVolume, maxVolume);
+    /// Using y=x^4 function as an approximation seems very natural and sufficient
+    /// For more info check: https://www.dr-lex.be/info-stuff/volumecontrols.html
+    volumeFactor = std::pow(1.0f * (vol / maxVolume), 4);
+    return AudioDevice::RetCode::Success;
+}
+void SAIAudioDevice::scaleOutputVolume(audio::Stream::Span &span)
+{
+    if (volumeFactor < 1.0) {
+        const auto data = reinterpret_cast<std::int16_t *>(span.data);
+        std::for_each(data, data + span.dataSize / 2, [this](auto &sample) { sample *= volumeFactor; });
+    }
+}
```

Comment:
is this unique for bell or is this impacting pure too?

Reply:
why limited to half ? data + span.dataSize / 2

Reply:
This won't affect Pure as it still uses hardware volume control. `audio::Stream::Span` stores its data as `uint8_t` but the audio codec uses 16-bit samples. We are iterating over the data using `uint16_t`s hence data range must be divided by 2.

Reply:
ok, so please add some comment, or event better name the new data size var

Reply:
Fixed